File: btc_f/lib/data/models/food_model.dart
class Food {
  final int id;
  final String name;
  final double price;
  final String description;
  final List<String> tagline;
  final String cuisine;
  final double rating;
  final int timeToDelivery;
  final String imageUrl;
  final int stockQuantity;
  final int calories;
  final String spicyLevel;
  final String vegNonveg;
  final String category;
  final String? review;

  Food({
    required this.id,
    required this.name,
    required this.price,
    required this.description,
    required this.tagline,
    required this.cuisine,
    required this.rating,
    required this.timeToDelivery,
    required this.imageUrl,
    required this.stockQuantity,
    required this.calories,
    required this.spicyLevel,
    required this.vegNonveg,
    required this.category,
    this.review,
  });

  factory Food.fromJson(Map<String, dynamic> json) {
    return Food(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      price: double.tryParse(json['price'].toString()) ?? 0.0,
      description: json['description'] ?? '',
      tagline:
          (json['tagline'] as List?)?.map((e) => e.toString()).toList() ?? [],
      cuisine: json['cuisine'] ?? '',
      rating: (json['rating'] as num?)?.toDouble() ?? 0.0,
      timeToDelivery: json['time_to_delivery'] ?? 0,
      imageUrl: json['image_url'] ?? '',
      stockQuantity: json['stock_quantity'] ?? 0,
      calories: json['calories'] ?? 0,
      spicyLevel: json['spicy_level'] ?? '',
      vegNonveg: json['veg_nonveg'] ?? '',
      category: json['category'] ?? '',
      review: json['review'],
    );
  }
}

File: Food-main/backend/models/foodModel.js
const { DataTypes } = require('sequelize');
const sequelize = require('../db');
const Review = require('../models/Review');

const Food = sequelize.define('Food', {
  name: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  price: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: false,
  },
  description: {
    type: DataTypes.TEXT,
  },
  restaurantId: {
  type: DataTypes.INTEGER,
  allowNull: false,
},
  tagline: {
    type: DataTypes.ARRAY(DataTypes.STRING), // ✅ tagline as array
    allowNull: true,
  },
  cuisine: {
    type: DataTypes.STRING,
  },
  rating: {
    type: DataTypes.FLOAT,
    defaultValue: 0.0,
  },
  time_to_delivery: {
    type: DataTypes.INTEGER,
  },
  image_url: {
    type: DataTypes.STRING,
  },
  stock_quantity: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  calories: {
    type: DataTypes.INTEGER,
  },
  spicy_level: {
    type: DataTypes.STRING,
  },
  veg_nonveg: {
    type: DataTypes.ENUM('Veg', 'Non-Veg'),
    allowNull: false,
  },
  category: {
    type: DataTypes.STRING,
  },
}, {
  tableName: 'foods',
  timestamps: true,
});

Food.hasMany(Review, { foreignKey: 'foodId' });


/** ✅ Static method to find similar products based on overlapping taglines */
Food.findSimilarProducts = async function(foodId) {
  const food = await Food.findByPk(foodId);
  if (!food || !food.tagline) return [];

  const similarFoods = await Food.findAll({
    where: {
      id: {
        [sequelize.Sequelize.Op.ne]: foodId // exclude current food
      },
      tagline: {
        [sequelize.Sequelize.Op.overlap]: food.tagline
      }
    }
  });

  return similarFoods;
};

module.exports = Food;
